plutus-contract/src/Wallet/Emulator/Wallet.hs

```diff
@@ -145,70 +150,167 @@ data PaymentArgs =
-    ( Member (Error WAPI.WalletAPIError) effs
-    )
-    => PaymentArgs
-    -> Payment
-    -> Eff effs Payment
-handleUpdatePaymentWithChange
-    PaymentArgs{availableFunds, ownPubKey, requestedValue}
-    Payment{paymentInputs=oldIns, paymentChangeOutput=changeOut} = do
-    let
-        -- These inputs have been already used, we won't touch them
-        usedFnds = Set.map txInRef oldIns
-        -- Optional, left over change. Replace a `Nothing` with a Value of 0.
-        oldChange = maybe (Ada.lovelaceValueOf 0) txOutValue changeOut
-        -- Available funds.
-        fnds   = Map.withoutKeys availableFunds usedFnds
-    if requestedValue `Value.leq` oldChange
-    then
-        -- If the requested value is covered by the change we only need to update
-        -- the remaining change.
-        pure Payment
-                { paymentInputs = oldIns
-                , paymentChangeOutput = mkChangeOutput ownPubKey (oldChange PlutusTx.- requestedValue)
-                }
-    else do
-        -- If the requested value is not covered by the change, then we need to
-        -- select new inputs, after deducting the oldChange from the value.
-        (spend, change) <-
-            selectCoin
-                (second (txOutValue . txOutTxOut) <$> Map.toList fnds)
-                (requestedValue PlutusTx.- oldChange)
-        let ins = Set.fromList (pubKeyTxIn . fst <$> spend)
-        pure Payment
-                { paymentInputs = Set.union oldIns ins
-                , paymentChangeOutput = mkChangeOutput ownPubKey change
-                }
-
 handleWallet ::
     ( Member NodeClientEffect effs
     , Member ChainIndexEffect effs
     , Member (State WalletState) effs
-    , Member (Error WAPI.WalletAPIError) effs
+    , Member (LogMsg TxBalanceMsg) effs
     )
     => WalletEffect ~> Eff effs
 handleWallet = \case
-    SubmitTxn tx -> W.publishTx tx
+    SubmitTxn tx -> do
+        logInfo $ SubmittingTx tx
+        W.publishTx tx
     OwnPubKey -> toPublicKey <$> gets _ownPrivateKey
-    UpdatePaymentWithChange vl pmt -> do
-        utxo <- W.watchedAddresses
-        ws <- get
-        let pubK   = toPublicKey (ws ^. ownPrivateKey)
-            args   = PaymentArgs
-                        { availableFunds = utxo ^. AM.fundsAt (ownAddress ws)
-                        , ownPubKey = pubK
-                        , requestedValue = vl
-                        }
-        handleUpdatePaymentWithChange args pmt
-    WalletSlot -> W.getClientSlot
-    OwnOutputs -> do
-        addr <- gets ownAddress
-        view (at addr . non mempty) <$> W.watchedAddresses
-    WalletAddSignature tx -> do
-        privKey <- gets _ownPrivateKey
-        pure (addSignature privKey tx)
+    BalanceTx utx -> runError $ do
+        logInfo $ BalancingUnbalancedTx utx
+        utxWithFees <- validateTxAndAddFees utx
+        -- balance to add fees
+        tx' <- handleBalanceTx (utx & U.tx . fee .~ (utxWithFees ^. U.tx . fee))
+        handleAddSignature tx'
```

Comment:
Move `logInfo $ FinishedBalancing tx'` to here, so the logs don't show the message twice?